tooling-docker/src/main/java/org/jboss/eap/qe/ts/common/docker/Docker.java

```diff
@@ -77,7 +79,31 @@ protected void start() throws Exception {
             }
         });
 
-        containerReadyCondition.waitUntilReady(containerReadyTimeout);
+        long startTime = System.currentTimeMillis();
+        while (!isContainerReady()) {
+            if (System.currentTimeMillis() - startTime > containerReadyTimeout) {
+                throw new Exception("Container was not ready in timeout : " + containerReadyTimeout + " ms");
+            }
+            // fail fast mechanism in case of malformed docker command, for example bad arguments, invalid format of port mapping, image version,...
+            if (!dockerRunProcess.isAlive() && dockerRunProcess.exitValue() != 0) {
+                throw new Exception("Starting of docker container using command: \"" + String.join(" ", cmd)
+                        + "\" failed. Check that provided command is correct.");
+            }
+        }
+    }
+
+    private boolean isContainerReady() throws Exception {
+        CompletableFuture<Boolean> condition = new CompletableFuture().supplyAsync(() -> containerReadyCondition.isReady());
+        try {
+            return condition.get(containerReadyTimeout, TimeUnit.MILLISECONDS);
+        } catch (TimeoutException ex) {
+            // in case condition hangs interrupt it so there are no zombie threads
+            condition.cancel(true);
+
+            throw new Exception("Provided ContainerReadyCondition.isReady() method took longer than containerReadyTimeout: "
```

Comment:
Could be providing a couple of strongly typed `Exception` sub-classes a good idea for refactoring? WDYT?

Reply:
+1 I was too lazy to do so :-) I'll fix it.